src/Logger.cc

```diff
@@ -0,0 +1,103 @@
+// SPDX-FileCopyrightText: Deutsches Elektronen-Synchrotron DESY, MSK, ChimeraTK Project <chimeratk-support@desy.de>
+// SPDX-License-Identifier: LGPL-3.0-or-later
+#include "Logger.h"
+
+#include <iostream>
+
+namespace ChimeraTK {
+
+  /********************************************************************************************************************/
+
+  Logger::~Logger() {
+    if(_mainLoopThread.joinable()) {
+      _mainLoopThread.interrupt();
+      _messageQueue.push_exception(std::make_exception_ptr(boost::thread_interrupted()));
+
+      // try joining the thread
+      while(!_mainLoopThread.try_join_for(boost::chrono::milliseconds(10))) {
+        // it may not suffice to send interrupt() once, as the exception might get
+        // overwritten in the queue, thus we repeat this until the thread was
+        // joined.
+        _messageQueue.push_exception(std::make_exception_ptr(boost::thread_interrupted()));
+      }
+    }
+    assert(!_mainLoopThread.joinable());
+  }
+
+  /********************************************************************************************************************/
+
+  Logger::StreamProxy Logger::getStream(Logger::Severity severity, std::string context) {
+    return {this, severity, std::move(context)};
+  }
+
+  /********************************************************************************************************************/
+
+  Logger::StreamProxy::StreamProxy(Logger* logger, Severity severity, std::string context)
+  : std::ostream(severity >= logger->_minSeverity ? &_buf : nullptr), _severity(severity), _context(std::move(context)),
+    _logger(*logger) {}
+
+  /********************************************************************************************************************/
+
+  Logger::StreamProxy::~StreamProxy() {
+    // Send out the log message
```

Comment:
I'm getting a linter warning about possibly escaping exception. I think d->semaphores[...].unlock() in future_queue.hpp, from push() is the origin. 
Just to be safe, you could catch the exception and maybe just print it to stderr.

Reply:
Right, I missed that (the little squiggle is too small to see...). I will make `log()` noexcept and catch the exception inside.